Sync users with OIDC provider

princemaple opened the pull request:
close #585 

This functionality is done in 2 parts:

1. Store OIDC connection (mainly the refresh token) when user login via an OIDC provider. **This can only be done when they sign in via the chose provider the first time. Existing users will have to revoke their access on the provider and re-login.**
2. Check users' OIDC connection status using the refresh token, periodically. If the user no longer has a valid status, disable their VPN session by timestamping their `disabled_at` so the rest of the system can pick it up and disable their wireguard interface.
3. Additionally, add the ability for admins to disable/re-enable specific users' permission to connect VPN.
![image](https://user-images.githubusercontent.com/1329716/170392937-96b0020c-49c2-4011-8a67-b71bd53df086.png)

![image](https://user-images.githubusercontent.com/1329716/170393099-680d0b84-b159-41bd-83f2-0f225d3a622a.png)

----

![image](https://user-images.githubusercontent.com/1329716/170393249-cca2208e-f5d7-436f-ae98-3d2a5da38aee.png)

----

![image](https://user-images.githubusercontent.com/1329716/170393268-138955f4-3289-45df-9c3b-ce45fcbb85df.png)

princemaple committed:
tmp [skip ci]

princemaple committed:
add connection [skip ci]

princemaple committed:
implement periodic refresh

princemaple committed:
format [skip ci]

princemaple committed:
no need to guard

princemaple committed:
make credo happy

jamilbk commented on apps/fz_http/lib/fz_http/oidc/refresher.ex:
```diff
@@ -0,0 +1,73 @@
+
+      user_id
+      |> Users.get_user!()
+      |> User.changeset(%{last_signed_in_at: ~U[1970-01-01 00:00:00Z]})
```
Hm, wonder if we should prevent disabling an `admin` user? Maybe if it's the last admin in the system?

princemaple replied:
Admins don't get disconnected either way, but if the point was to not change their last logged in time, the point is valid.

princemaple replied:
> Ah also, we should have some mechanism for an administrator to re-enable an accidentally-disabled user... maybe a UI form?

They could add the user back to their OIDC provider? If you meant not using an OIDC login, I guess they can remove the user's OIDC connection - then they will not be checked, hence won't be disabled.

jamilbk commented on apps/fz_http/lib/fz_http/oidc/refresher.ex:
```diff
@@ -0,0 +1,73 @@
+
+      user_id
+      |> Users.get_user!()
+      |> User.changeset(%{last_signed_in_at: ~U[1970-01-01 00:00:00Z]})
```
Instead of changing `last_signed_in_at` here, I think we'll probably have to add a `disabled_reason` string on `users` to track disabled Users and check for this in `Devices.to_peer_list/0` since we allow never-signed-in Users to connect to VPN if the adminstrator has not turned on the `Require users to authenticate for VPN` setting.

`disabled_reason` could be set to `oidc_refresh_error` for now. Later down the road we could add more, e.g. `too_many_password_attempts` or `disabled_by_admin` (later down the road).

jamilbk commented on apps/fz_http/lib/fz_http/oidc/refresh_manager.ex:
```diff
@@ -0,0 +1,44 @@
+    schedule()
+
+    User
+    |> Repo.all()
```
Maybe this could be optimized to only fetch users that have an associated `oidc_connection`?

princemaple replied:
Yep, I omitted a lot of optimization, intentionally. As I wasn't sure, what we will do eventually, like having a manager per site, or per user, or per site per user, or per site per provider, etc.

jamilbk replied:
Sounds good.

princemaple committed:
remove filter

jamilbk commented on apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex:
```diff
@@ -67,6 +67,16 @@ defmodule FzHttpWeb.AuthController do
          {:ok, claims} <- openid_connect.verify(provider, tokens["id_token"]) do
       case UserFromAuth.find_or_create(provider, claims) do
         {:ok, user} ->
+          # only first-time connect will include refresh token
```
Hmm thinking through this, if Firezone is offline for an extended time, will that disable all OIDC users permanently? Or will it just force them to re-authenticate at the next sign-in?

jamilbk reviewed (commented):
Nice approach! The only gotcha I think we need to be careful of is disabling users if Firezone is offline or has no connectivity for an extended period.

Ah also, we should have some mechanism for an administrator to re-enable an accidentally-disabled user... maybe a UI form?

princemaple committed:
updates

princemaple committed:
extract component

princemaple committed:
prevent changing admin's connection permission

princemaple committed:
add to show as well

princemaple committed:
show connections

princemaple committed:
hide oidc connections section if nothing is there

princemaple committed:
add test

princemaple committed:
leave site id out for now

may not need it if we create different users for different sites, even
for the same email

princemaple committed:
move delay randomization to manager

princemaple committed:
add test

jamilbk commented on apps/fz_http/lib/fz_http/oidc/refresher.ex:
```diff
@@ -0,0 +1,72 @@
+    })
+
+    with %{error: _} <- refresh_response do
+      Logger.info("Disabling user\##{user_id}...")
```
May want to state why here and log the email instead of user_id? , e.g.

```elixir
Logger.info("Disabling user #{user.email} due to OIDC token refresh failure...")
```

jamilbk commented on apps/fz_http/lib/fz_http/oidc/refresher.ex:
```diff
@@ -0,0 +1,72 @@
+
+    with %{error: _} <- refresh_response do
+      Logger.info("Disabling user\##{user_id}...")
+
```
Could you add a new [telemetry function](https://github.com/firezone/firezone/blob/master/apps/fz_http/lib/fz_http/telemetry.ex) and call it here to track events like this?

```elixir
def disable_user(user, reason = "oidc refresh failure") do
      telemetry_module().capture(
      "disable_user",
      common_fields() ++
        [
           user_email_hash: hash(user.email),
           reason: reason
        ]
    )
end
```

jamilbk commented on apps/fz_http/lib/fz_http_web/live/user_live/allowed_to_connect_component.ex:
```diff
@@ -0,0 +1,28 @@
+
+  @impl Phoenix.LiveComponent
+  def handle_event("toggle_allowed_to_connect", params, socket) do
+    user =
```
Could you call the new telemetry function here too?

jamilbk commented on apps/fz_http/lib/fz_http_web/templates/shared/user_details.html.heex:
```diff
@@ -10,6 +10,14 @@
       <td><%= @user.role %></td>
     </tr>
 
+    <tr>
```
Hmmm, maybe we should make this non-editable? I wonder if Admins will need to do this en-masse, we might not want to make it this easy for admins to accidentally enable/disable VPN connections.

jamilbk commented on apps/fz_http/lib/fz_http_web/live/user_live/index.html.heex:
```diff
@@ -33,6 +34,9 @@
             <%= live_patch(user.email, to: Routes.user_show_path(@socket, :show, user)) %>
           </td>
           <td id={"user-#{user.id}-device-count"}><%= user.device_count %></td>
+          <td>
```
What do you think about moving this down to the `Danger Zone` section below? It feels like enabling / disabling a user's VPN should come with a bit more thought.

jamilbk commented on apps/fz_http/lib/fz_http_web/live/user_live/index.html.heex:
```diff
@@ -20,6 +20,7 @@
         <tr>
           <th>Email</th>
           <th>Devices</th>
+          <th>Can Connect</th>
```
See comment Re: moving this down to the section below.

jamilbk commented on apps/fz_http/lib/fz_http_web/live/user_live/allowed_to_connect_component.ex:
```diff
@@ -0,0 +1,28 @@
+  @impl Phoenix.LiveComponent
+  def render(assigns) do
+    ~H"""
+    <input type="checkbox" phx-target={@myself} disabled={@user.role == :admin}
```
Thoughts on adding `data-confirm="Are you sure? This may affect this user's internet connectivity."`

jamilbk reviewed (changes requested):
Looks good, just a few suggestions. Also, a couple design improvements I think we should consider:

* What do you think of refactoring `allowed_to_connect:boolean` to `disabled_at:datetime` field?
* Should probably re-allow VPN sessions after a user successfully authenticates in `Authentication.sign_in/2`
* Should probably disconnect user_sockets related to the user when they get disabled... prevents an admin from Re-enabling their own VPN connection from a connected browser tab

princemaple committed:
address comments

princemaple committed:
re-enable user when sign-ed via oidc

princemaple committed:
fix test

princemaple committed:
remove unused alias

princemaple committed:
fix tests

princemaple committed:
update tests

princemaple committed:
fix re-enable

princemaple committed:
add tests

princemaple committed:
Update vpn_connection_component_test.exs

princemaple committed:
revert formats

princemaple committed:
tests

princemaple committed:
remove alias

princemaple committed:
remove special case for admins

princemaple committed:
remove auto re-enable

jamilbk commented on apps/fz_http/test/support/test_helpers.ex:
```diff
@@ -146,7 +146,9 @@ defmodule FzHttp.TestHelpers do
   end
 
-  def create_users(%{count: count}) do
+  def create_users(opts) do
```
Thanks :-)

jamilbk reviewed (approved):
LGTM! This is awesome. Nice work!
